// gherkin/dist/src/TokenFormatterBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TokenFormatterBuilder = /** @class */ (function () {
    function TokenFormatterBuilder() {
        this.tokensText = '';
    }
    TokenFormatterBuilder.prototype.reset = function () {
        this.tokensText = '';
    };
    TokenFormatterBuilder.prototype.startRule = function (ruleType) {
        // no-op
    };
    TokenFormatterBuilder.prototype.endRule = function (ruleType) {
        // no-op
    };
    TokenFormatterBuilder.prototype.build = function (token) {
        this.tokensText += this.formatToken(token) + '\n';
    };
    TokenFormatterBuilder.prototype.getResult = function () {
        return this.tokensText;
    };
    TokenFormatterBuilder.prototype.formatToken = function (token) {
        if (token.isEof) {
            return 'EOF';
        }
        return ('(' +
            token.location.line +
            ':' +
            token.location.column +
            ')' +
            token.matchedType +
            ':' +
            (typeof token.matchedKeyword === 'string' ? token.matchedKeyword : '') +
            '/' +
            (typeof token.matchedText === 'string' ? token.matchedText : '') +
            '/' +
            token.matchedItems
                .map(function (i) {
                return i.column + ':' + i.text;
            })
                .join(','));
    };
    return TokenFormatterBuilder;
}());
exports.default = TokenFormatterBuilder;
